pkg/versioning/properties.go

```diff
@@ -0,0 +1,90 @@
+package versioning
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/magiconair/properties"
+	"github.com/pkg/errors"
+)
+
+// PropertiesFile defines an artifact using a properties file for versioning
+type PropertiesFile struct {
+	path             string
+	content          *properties.Properties
+	versioningScheme string
+	versionField     string
+	writeFile        func(string, []byte, os.FileMode) error
+}
+
+func (p *PropertiesFile) init() error {
+	if len(p.versionField) == 0 {
+		p.versionField = "version"
+	}
+	if p.writeFile == nil {
+		p.writeFile = ioutil.WriteFile
+	}
+	if p.content == nil {
+		var err error
+		p.content, err = properties.LoadFile(p.path, properties.UTF8)
+		if err != nil {
+			return errors.Wrapf(err, "failed to load file %v", p.path)
+		}
+	}
+	return nil
+}
+
+// VersioningScheme returns the relevant versioning scheme
+func (p *PropertiesFile) VersioningScheme() string {
+	if len(p.versioningScheme) == 0 {
+		return "semver2"
+	}
+	return p.versioningScheme
+}
+
+// GetVersion returns the current version of the artifact with a ini-file-based build descriptor
+func (p *PropertiesFile) GetVersion() (string, error) {
+	if p.content == nil {
+		err := p.init()
+		if err != nil {
+			return "", err
+		}
+	}
```

Comment:
There seems to be no benefit of checking p.content. I would expect GetVersion() to be called only once, and in that case it actually adds (negligible) overhead.

Reply:
consistently implemented as in other versioning implementation, e.g. `INIfile`.
We cannot assume that get and set is executed in order.

Reply:
Sorry for being unclear. What I mean is that p.init() already contains a check for `p.content == nil`. So I would remove the outer check and just call `p.init()` unconditionally.

Reply:
ah, ok - now I get it :-)
makes sense

Reply:
is changed ...